<!DOCTYPE html>
<html>
    <head>
        <title>GeLL</title>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
        <link rel="stylesheet" type="text/css" href="style.css" />
        <script type="text/javascript" src="script.js"></script>
    </head>
    <body>
        <div style="width: 800px; margin: 0px auto">
            <H1>GeLL</H1>
            <p>GeLL is a <b>Ge</b>neral <b>L</b>ikelihood <b>L</b>ibrary for 
            use in phylogenetics.  It is intended to provide classes that facilitate
            investigating new models and techniques.  It is not intended for everyday
            use in phylogenetic studies as it is optimised for easy use of new models
            rather than speed.</p>

            <p>GeLL achieves this versatility by allowing models to be defined using
            an array of Strings where each string represents an equation.  This allows
            new models to be defined quickly and easily.  The driver executable included
            allows many different models to be run simply by changing the text file
            defining the model. </p>
            
            <div class="collapsetop">Packages
                <span class="show" onclick="show('quickstart')" id="quickstartS">Show</span>
                <span class="hide" onclick="hide('quickstart')" id="quickstartH">Hide</span>
            </div>
            <div class="collapseinner" id="quickstart">
                A quick start guide, including instalation instructions, is avaliable.
                Click here to get it.  It is also included in every download package.
                (quickstart.pdf).
            </div>
            
            <div class="collapsetop">Capabilities
                <span class="show" onclick="show('cap')" id="capS">Show</span>
                <span class="hide" onclick="hide('cap')" id="capH">Hide</span>
            </div>
            <div class="collapseinner" id="cap" style="display:block">
                GeLL has the following capabilities:
                <ul >
                <li>
                    <b>Likelihood</b> - Using the method of Felsenstein 1981<sup>[1]</sup>.  Unobserved
                    data can also be accounted for, per Felsenstein 1992<sup>[2]</sup>.  Confidence intervals
                    for a parameter can also be calculated as discussed in Felsenstein 2003<sup>[3]</sup>.
                </li>
                <li>
                    <b>Ancestral Reconstruction</b>
                    <ul>
                        <li>
                            <i>Marginal Reconstruction</i> - Using the basic method of Yang, 
                                    Kurma and Nei 1995<sup>[4]</sup>.
                        </li>
                        <li>
                            <i>Joint Reconstruction</i> - Using the methods of Pupko et al 2000<sup>[5]</sup> and
                            Pupko et al 2002<sup>[6]</sup>.
                        </li>
                    </ul>
                </li>
                <li>
                    <b>Simulated Data</b> - Using the standard methods as described, for example,
                    in Yang 2006<sup>[7]</sup>.
                </li>
                <li>
                    <b>Non-reversible Processes</b> - Using the method described by Boussau and Gouy 2006<sup>[8]</sup>.
                </li>
            </ul>
            </div>

            <div class="collapsetop">Packages
                <span class="show" onclick="show('release')" id="releaseS">Show</span>
                <span class="hide" onclick="hide('release')" id="releaseH">Hide</span>
            </div>
            <div class="collapseinner" id="release">
                <p>GeLL comes in three different packages:</p>
                <ul>
                    <li><b>Basic Package</b> - This includes just the JAR file
                        and the javadoc documentation</li>
                    <li><b>Source Package</b> - This additionally includes the
                        source code</li>
                    <li><b>Test Package</b> - This additionally includes both
                        the source code and test classes
                </ul>
            </div>
            
            <div class="collapsetop">Documentation
                <span class="show" onclick="show('doc')" id="docS">Show</span>
                <span class="hide" onclick="hide('doc')" id="docH">Hide</span>
            </div>
            <div class="collapseinner" id="doc">
                <p>GeLL comes with full javadoc documentation in the javadoc
                directory <a href="javadoc/index.html">here</a>.</p>
                
                <p>GeLL includes a driver that should be capable of doing
                most basic computations.  This is documented in the driver
                documentation <a href="Driver.html">here</a>.</p>
                
                <p>The GeLL packages also include an example program.  As well
                as the source code file an annotated version of the source is
                <a href="Example.html">here</a>.  If you are thinking of writing
                your own program using the GeLL library it is recommended you
                take a look at this example.</p>
                
                <p>The core capability of GeLL is the ability to define and use 
                new models.  As such how models are defined in covered in more
                details <a href="Model.html">here</a>.  Also included is a
                more in-depth discussion of how parameters are defined and
                constrained. </p>
                
                <p>A description of the test cases is included <a href="Tests.html">
                here</a> in the test package.</p>
            </div>
            
            <div class="collapsetop">Tips & Tricks
                <span class="show" onclick="show('tips')" id="tipsS">Show</span>
                <span class="hide" onclick="hide('tips')" id="tipsH">Hide</span>
            </div>
            <div class="collapseinner" id="tips">
                <ul>
                    <li><b>Gaps</b><br/>
                        Gaps can be dealt with in two ways.  They can be treated
                        as an extra state in which case the extra state should simply
                        be added to the model.  To sum across all states, like many
                        models do, simply define the gap character as being an ambiguous
                        character that could be any one of the states.</li>
                </ul>
            </div>
            
            <div class="collapsetop">Citation
                <span class="show" onclick="show('citation')" id="citationS">Show</span>
                <span class="hide" onclick="hide('citation')" id="citationH">Hide</span>
            </div>
            <div class="collapseinner" id="citation">
                Money, D. and Whelan, S. (2012) GeLL: Generalized Likelihood Library
                (http://phylo.bio.ku.edu/GeLL)
            </div>
            
            <div class="collapsetop">Known Issues
                <span class="show" onclick="show('issues')" id="issuesS">Show</span>
                <span class="hide" onclick="hide('issues')" id="issuesH">Hide</span>
            </div>
            <div class="collapseinner" id="issues">
                <ul>
                    <li><b>Eigendecomposition</b><br/>
                    Eigendecompositions may not always converge and many not throw
                    an error.  The eigendecomposition code is taken from
                    <a href="http://math.nist.gov/javanumerics/jama/">JAMA</a> and
                    has not been debugged.  This seems to especially be the case for
                    gene family size models.</li>
                    <li><b>Parameters (1)</b><br/>
                    There is an issue where using the same Parameters object in two or more
                    calls to different Calculator objects will cause an error.  This
                    can be solved by passing a cloned copy of the parameters in the second
                    <li><b>Parameters (2)</b><br/>
                    The way branch lengths are treated as parameters is a little odd
                    and could well cause confusion.  It is intended to overhaul
                    this in a future version.
                    <li><b>Exceptions</b><br/>
                    Some exceptions are thrown in circumstances where it is known
                    that the exception should not occur, normally because we know
                    that the calling class has already checked (either implicitly
                    or explicitly) that the exception can't occur.  It is intended
                    to fix this in a later release.</li>
                    <li><b>Threaded Computation - Likelihood</b><br/>
                    Currently threaded computation is always used for likelihood
                    computation.  It is known that this can be slower for small
                    trees and small state-space models.  However, it has not been
                    yet been determined under what conditions threaded calculation
                    is slower so, for the moment, it is always used.</li>
                    <li><b>Threaded Computation - Ancestral reconstruction</b><br/>
                    At the moment ancestral reconstruction is not threaded.</li>
                    <li><b>Optimising No Parameters</b><br/>
                    Trying to optimise a parameters object with no optimisable
                    parameters (i.e. they are all fixed values) can result in an
                    infinite loop.</li>
                </ul>
            </div>
            
            <div class="collapsetop">Release Notes
                <span class="show" onclick="show('notes')" id="notesS">Show</span>
                <span class="hide" onclick="hide('notes')" id="notesH">Hide</span>
            </div>
            <div class="collapseinner" id="notes">
                <b>1.0</b> <span class="reldate">2 February 2012</span>
                <ul class="changes">
                    <li>First release</li>
                </ul>
                <b>1.1</b> <span class="reldate">4 April 2012</span>
                <ul class="changes">
                    <li><i>Not publicly release due to memory problems</i></li>
                    <li>Added support for model hypothesis testing</li>
                    <li>Allow sites to be in different class</li>
                </ul>
                <b>1.2</b> <span class="reldate">23 April 2012</span>
                <ul class="changes">
                    <li><i>Not (quite) backwards compatible although only minimal changes should be necessary</i></li>
                    <li>Changes to internal storage structures to sort memory problems</li>
                    <li>Large amounts of code optimisation</li>
                </ul>
                <b>1.3</b> <span class="reldate">14 August 2012</span>
                <ul class="changes">
                    <li><i>Not (quite) backwards compatible although only minimal changes should be necessary</i></li>
                    <li>Added Conjugate Gradient search</li>
                    <li>Added an ID field to Site</li>
                    <li>Added split support to Tree and also the ability to calculate
                        various distances between trees (RF, weighted RF and branch score).</li>
                    <li>Minor Changes:
                    <ul class="minor">
                        <li>Optimiser don't store node likelihood until after optimisation
                            is complete to save memory.</li>
                        <li>Attempt to make calculating distributions using the
                            "repeated" method more stable.</li>
                        <li>Fixed bugged in test for quasi-stationary distribution
                            convergence.</li>
                        <li>Parameters toString method now only outputs estimated
                            parameters by default.</li>
                        <li>Ability to write sequence alignments in Phylip Alignments
                            in PAML format, i.e. taxa names have a maximum of 15 characters.
                            This is buggy as names are simply truncated and duplicated names
                            are not checked for.</li>
                        <li>Fix to PossibleSettings which meant settings in no group did
                            not work properly.</li>
                        <li>Parameter now throws an error if an attempt is made to set it
                            to an invalid value.</li>
                        <li>Additional error checking in likelihood calculations to check
                            for probabilities greater than one and other errors.</li>
                        <li>Various updates to throw statements caused by the above.</li>
                        <li>Moved the storing of whether matrices need to be recalculated
                            from Parameters to RateCategory.  This is more inituative as the
                            what parameters are in the rate category decides when changes to
                            parameters needs matrix recalculations. Should, hopefully, be largely 
                            transparent.</li>
                    </ul>
                    </li>
                </ul>
                <b>2.0</b> <span class="reldate">14 April 2014</span>
                <ul class="changes">
                    <li><i>Not (quite) backwards compatible although only minimal changes should be necessary</i></li>
                    <li>Added the ability to work with very small likelihoods.  This
                        is optional as it results in a significant slow down.  There
                        is a slight performance decrease even when the small option
                        is not being used. </li>
                    <li>Added the ability to use the FitzJohn (2009) method as the
                        root distribution.</li>
                    <li>Major changes to how calculators etc are structured to
                        allow more unusual ways of calculating likelihoods.</li>
                    <li>Removed code to do with Constraints as it was implemented
                        in a very odd way and was no longer needed. </li>
                    <li>Added BDIE (Birth-Death-Innovation-Extinction) and
                        Birth-Death (with no zero state) models to the duplication
                        model factory.</li>
                    <li>Minor Changes:
                    <ul class="minor">
                        <li>Removed ArrayMap as it was not providing any speed
                            increase so was just complicating the code.</li>
                        <li>Removed ToDoubleHashMap, again as it was complicating
                            the code.</li>
                        <li>Removed deprecated SequenceAlignment.  PhylipAlignment
                            should be used instead.</li>
                        <li>Removed unused MapUtils.</li>
                        <li>Added an additional example executable that implements
                            the model of chromosome evolution of Mayrose et al 
                            (2009).  This example shows how to implement a new model.</li>
                        <li>DNAModel factory now have two options to create models.
                            One functions takes in and updates the parameters list
                            with relevant parameters whereas the other just creates
                            the model.</li>
                        <li>Code now automatically checks whether a parameter is
                            in the matrix so it no longer part of the Parameter.
                            Relevant constructors removed.</li>
                        <li>Added an ability to generate a random unrooted tree.</li>
                    </ul>
                    </li>
                </ul>
            </div>
            
            <div class="collapsetop">License
                <span class="show" onclick="show('license')" id="licenseS">Show</span>
                <span class="hide" onclick="hide('license')" id="licenseH">Hide</span>
            </div>
            <div class="collapseinner" id="license">
                <p>GeLL is avaliable under <a href="gpl.txt">GPL v3</a>.</p>
                <p>The <code>EigenvalueDecomposition</code> and <code>LUDecomposition</code>
                classes are from <a href="http://math.nist.gov/javanumerics/jama/">JAMA</a>
                and as such are in the public domain.</p>
                <p>The test package contains both executables and example datasets from 
                <a href="http://abacus.gene.ucl.ac.uk/software/paml.html">PAML</a>.  This
                example dataset is also used for the example execution using <code>
                Example.java</code>.  The applicable license for these components is as
                follows:</p>
                <div style="width: 95%; margin: 0px 0px 0px auto">
                  © Copyright 1993-2008 by Ziheng Yang<br/>
                  The software package is provided "as is" without warranty of any kind. 
                  In no event shall the author or his employer be held responsible for any 
                  damage resulting from the use of this software, including but not limited 
                  to the frustration that you may experience in using the package. The program 
                  package, including source codes, example data sets, executables, and this 
                  documentation, is distributed free of charge for academic use only. 
                  Permission is granted to copy and use programs in the package provided no 
                  fee is charged for it and provided that this copyright notice is not removed.
                </div>
            </div>
            
            <div class="collapsetop">Contact
                <span class="show" onclick="show('contact')" id="contactS">Show</span>
                <span class="hide" onclick="hide('contact')" id="contactH">Hide</span>
            </div>
            <div class="collapseinner" id="contact">
                <p>The latest stable release of the library is available at 
                    <a href="http://phylo.bio.ku.edu/GeLL">http://phylo.bio.ku.edu/GeLL</a>,
                    while the latest developmental version is available at 
                    <a href="http://github.com/danielmoney/GeLL">http://github.com/danielmoney/GeLL</a>.</p>
                <p>The author is contactable at <a href="mailto:daniel_money@ku.edu">
                daniel.money@dal.ca</a></p>
            </div>
            
            <div class="collapsetop">References
                <span class="show" onclick="show('ref')" id="refS">Show</span>
                <span class="hide" onclick="hide('ref')" id="refH">Hide</span>
            </div>
            <div class="collapseinner" id="ref">
                <ol style="font-size:small">
                    <li>Felsenstein, J. 1981. Evolutionary trees from DNA sequences: A maximum likelihood
                        approach. <i>Journal of Molecular Evolution</i>, 17:368–376.</li>
                    <li>Felsenstein, J. 1992. Phylogenies from restriction sites: A Maximum-Likelihood
                        approach. <i>Evolution</i>, 46:159–173.</li>
                    <li>Felsenstein, J. 2003. <i>Inferring phylogenies</i>. Sinauer Associates Sunderland, Mass., USA.</li>
                    <li>Yang, Z., S. Kumar, and M. Nei. 1995. A new method of inference of ancestral nucleotide
                        and amino acid sequences. <i>Genetics</i>, 141:1641–1650.</li>
                    <li>Pupko, T., I. Pe, R. Shamir, and D. Graur. 2000. A fast algorithm for joint reconstruction
                        of ancestral amino acid sequences. <i>Molecular Biology and Evolution</i>, 17:890–896.</li>
                    <li>Pupko, T., I. Pe’er, M. Hasegawa, D. Graur, and N. Friedman. 2002. A branchand-
                        bound algorithm for the inference of ancestral amino-acid sequences when the
                        replacement rate varies among sites: Application to the evolution of five gene families.
                        <i>Bioinformatics</i>, 18:1116–1123.</li>
                    <li>Yang, Z. 2006. <i>Computational molecular evolution</i>. Oxford University Press, USA.</li>
                    <li>Boussau, B. and Gouy, M. 2006 Efficient Likelihood Computations with Nonreversible 
                        Models of Evolution.  <i>Systematic Biology</i>, 55:756-768. </li>
                </ol>
            </div>
        </div>
        <script type="text/javascript">                
             hide('release'); hide('doc'); hide('tips'); hide('citation'); hide('issues');
             hide('notes'); hide('license'); hide('contact'); hide('ref'); hide('cap');
             show('quickstart');
        </script>
    </body>
</html>
